### ISSUE TITLE
How to use vispy with higher version GLSL?

### ISSUE THREAD

**Issue by wangg12** [vispy/vispy#1630]
I want to use vispy with higher version GLSL, for example version 450 like this:
```
frag = '''
#version 450 core

in vec3 v_color;
in vec3 v_view;
in vec3 ViewDir;

in vec3 v_L;
in vec3 v_normal;

layout (location = 0) uniform float u_light_ambient_w;
layout (location = 2) uniform float u_light_diffuse_w;
layout (location = 3) uniform float u_light_specular_w;

layout (location = 0) out vec3 rgb;
layout (location = 1) out float depth;
layout (location = 2) out vec3 rgb_normals;

void main(void) {

    vec3 Normal = normalize(v_normal);
    vec3 LightDir = normalize(v_L);
    vec3 ViewDir = normalize(v_view);

    vec3 diffuse = max(dot(Normal, LightDir), 0.0) * v_color;
    vec3 R = reflect(-LightDir, Normal);
    vec3 specular = max(dot(R, ViewDir), 0.0) * v_color;


    rgb = vec3(u_light_ambient_w  * v_color +
               u_light_diffuse_w  * diffuse +
               u_light_specular_w * specular);

    if(rgb.x > 1.0) rgb.x = 1.0;
    if(rgb.y > 1.0) rgb.y = 1.0;
    if(rgb.z > 1.0) rgb.z = 1.0;
    rgb_normals = vec3(Normal * 0.5 + 0.5); // transforms from [-1,1] to [0,1]
    depth = v_view.z;
}
'''
```

However I can't figure out how to bind this kind of GLSL code with vispy, any ideas?

**Comment by djhoese** [comment]
What have you tried? This should work, but is highly dependent on the backend you are using, what your OS allows, and what your GPU/drivers are capable of.

For example, on OSX I had trouble getting the newly added geometry shaders (not released yet) to work. On OSX, at least from what I remember, you couldn't choose a "core" GL profile with the newest versions of GL without converting all GL logic to be 4.0+. The way vispy has some of it's low-level internal logic did not seem to fit the requirements of GL 4.0+ so it failed to work.

For the backend, some backends require that you explicitly state/set that you want to use the GL core profile to get some higher versions of the GL; by default they will use the compatibility profile. Vispy does not currently provide a way, as far as I know, to easily configure backends to change profile but I wouldn't be against it.

Bottom line, *most* things should work so I'm curious what errors you get. Depending on the features you hope to use and what GL profiles they use (and what OS you want to use them on), this may not be possible with current vispy. We would have to drop GL 2.0 or ES support based on what I saw the last time I experimented with this.

@rougier @campagnola If you have any experience with this in glumpy or vispy, your opinion is welcome.

**Comment by wangg12** [comment]
@djhoese I just don't know how to bind the attributes and get the outputs in vispy. I can only find examples which use the old `version 120` code. The GLSL code provided above works fine with a custom python OpenGL program but I wonder how to achieve that in vispy.

**Comment by djhoese** [comment]
Ah ok, so you've created a custom Visual (or used gloo code) and created the shaders but can't figure out how to set the values for things like `u_light_ambient_w`?

**Comment by wangg12** [comment]
Yes. And I don't know how to fetch the output values. In the old version, the output color is obtained from `gl_FragColor`, but in newer version, we should be able to fetch any custom values like color, depth, normals.

**Comment by rougier** [comment]
On OSX, you're quite limited in any case. But the hardcore problem is that we targeted GL ES 2.0 from the beginning and I'm not even sure our GLSL parser can handle keyword like `in`or `out`.

**Comment by djhoese** [comment]
The closest thing you'll find to a working example is the geometry shader example here: https://github.com/vispy/vispy/blob/master/examples/basics/gloo/geometry_shader.py

Looking at the code that parses the shader code we can handle `in` and `out` but not `layout`. You'd have to edit the regular expression and handling of it here: https://github.com/vispy/vispy/blob/master/vispy/gloo/program.py#L242-L255

**Comment by rougier** [comment]
Oh great.

**Comment by djhoese** [comment]
@wangg12 Something that maybe I didn't mention. If you aren't using current master then please switch to it. There was a lot of work with shaders done by @campagnola to get geometry shaders working that may help you here.

**Comment by larsoner** [comment]
Closing since there are some workarounds mentioned above
